### PR TITLE
Project Casper

### PR DESCRIPTION
**Pull request by cbwebdevelopment**
See https://trello.com/c/Jdwx1isM for details.

Related PRs:
tidepool-org/tideline#346
tidepool-org/viz#110

### CODE REVIEW

**Comment by cbwebdevelopment** [comment]
[[security] Project Casper](https://trello.com/c/Jdwx1isM/424-security-project-casper)

**Comment by cbwebdevelopment** [comment]
@pazaan Could you please review the latest changes?  I've changed the script-src to use nonces (and I'm applying them via a middleware on a per-request basis) instead of the whitelist approach we initially discussed, as that seems to be the current best practice.

References: 
https://csp.withgoogle.com/docs/strict-csp.html
https://websec.be/blog/cspstrictdynamic/

File: server.js [hunk 1]
```diff
@@ -1,26 +1,78 @@
...
+var nonceMiddleware = function (req, res, next) {
+  // Cache static html file to avoid reading it from the filesystem on each request
+  if (!global.html) {
+    console.log('Caching static HTML');
+    global.html = fs.readFileSync(staticDir + '/index.html', 'utf8');
+  }
+
+  // Set a unique nonce for each request
+  res.locals.nonce = crypto.randomBytes(16).toString('base64');
+  res.locals.htmlWithNonces = global.html.replace(/<(script)/g, '<$1 nonce="' + res.locals.nonce + '"');
```
Comment (pazaan) [inline server.js]: Recommend using template literals here for readability: 
`'<$1 nonce="' + res.locals.nonce + '"' ` -> `` `<$1 nonce="${res.locals.nonce}"` ``

File: server.js [hunk 2]
```diff
@@ -1,26 +1,78 @@
...
 var buildDir = 'dist';
 
 var app = express();
-var helmet = require('helmet');
+
+var nonceMiddleware = function (req, res, next) {
+  // Cache static html file to avoid reading it from the filesystem on each request
+  if (!global.html) {
+    console.log('Caching static HTML');
+    global.html = fs.readFileSync(staticDir + '/index.html', 'utf8');
```
Comment (pazaan) [inline server.js]: Recommend template literals for readability:
`fs.readFileSync(staticDir + '/index.html', 'utf8');` -> `` fs.readFileSync(`${staticDir}/index.html`, 'utf8'); ``

File: server.js [hunk 3]
```diff
@@ -1,26 +1,78 @@
...
 app.use(helmet());
+app.use(nonceMiddleware, helmet.contentSecurityPolicy({
+  directives: {
+    defaultSrc: ["'none'"],
+    baseUri: ["'none'"],
+    scriptSrc: [
+      "'self'",
+      "'strict-dynamic'",
+      function (req, res) {
+        return "'nonce-" + res.locals.nonce + "'";
```
Comment (pazaan) [inline server.js]: Recommend template literal.

File: .eslintrc [hunk 4]
```diff
@@ -24,7 +24,9 @@
   },
   // To give you an idea how to override rule options:
   "rules": {
-    "quotes": [2, "single"],
+    "quotes": [2, "single", {
+      "avoidEscape": true
```
Comment (pazaan) [inline .eslintrc]: If you end up using template literals, you can get rid of this.

**Comment by cbwebdevelopment** [comment]
@pazaan I'll change the whole file over to es6 (including template literals).   A lot of blip is written in es5 styles, and we sometimes convert them over as we go, sometimes not.  

We usually either convert the whole file over at once, or, for small changes, leave it as is and keep new code consistent with the rest of the file.

It won't take long at all, so I agree it might as well be done now.

**Comment by cbwebdevelopment** [comment]
@pazaan Updated to ES6.

**Pull request merged**